docs/react-v9/contributing/rfcs/shared/build-system/stop-styles-transforms.md

```diff
@@ -0,0 +1,208 @@
+```
+
+The single reliable solution is to prefix styles with a unique identifier. This prevents CSS rules from colliding, for example:
+
+```html
+<!-- PROPOSAL -->
+
+<!-- main bundle -->
+<style>
+  .main-order0 {
+    padding: 10px;
+  }
+  .main-order1 {
+    padding-left: 5px;
+  }
+</style>
+
+<!-- CDN bundle -->
+<style>
+  .order0 {
+    padding: 10px;
+  }
+</style>
+
+<!-- HTML -->
+
+<!--
+  ✅ Now the "div" below will have "padding-left: 5px" as CSS rules don't clash anymore.
+  -->
+
+<div class="main-order0 main-order1"></div>
+```
+
+The proposed idea is to integrate prefix as a part of classes generation in Griffel as a salt for hashing (see [`hashClassName`](https://github.com/microsoft/griffel/blob/e3bf466e36464585ecff0ca3492760f6ada76dcc/packages/core/src/runtime/utils/hashClassName.ts#L23-L32) function):
+
+- 👍 Class length won't change
+- 👍 Apps can easily _prefix_ classes with a unique identifier like a project name
+
+## Problem statement
+
+Integrating prefixing into Griffel is relatively straightforward. However, we currently undergo a pre-processing step in `@fluentui/react-components` that modifies JavaScript code.
+
+<details>
+  <summary>Original code</summary>
+
+```ts
+// packages/react-components/react-menu/src/components/MenuDivider/useMenuDividerStyles.styles.ts
+// 📝 output is simpfied
+
+const useStyles = makeStyles({
+  root: {
+    ...shorthands.margin('4px', '-5px', '4px', '-5px'),
+  },
+});
+```
+
+</details>
+
+<details>
+  <summary>Pre-processed code</summary>
+
+```ts
+// @fluentui/react-menu/lib/components/MenuDivider/useMenuDividerStyles.styles.js
+// 📝 output is simpfied
+
+const useStyles = /*#__PURE__*/ __styles(
+  {
+    root: {
+      B6of3ja: 'fvjh0tl',
+      t21cq0: ['f1rnx978', 'f1q7jvqi'],
+    },
+  },
+  {
+    d: ['.fvjh0tl{margin-top:4px;}', '.f1rnx978{margin-right:-5px;}', '.f1q7jvqi{margin-left:-5px;}'],
+  },
+);
+```
+
+</details>
+
+Addressing prefixing for pre-processed styles presents a challenge, as they have already been transformed into CSS rules and classes:
+
+- 👎 **Webpack loader** (i.e. build time) - Will make loaders more complex and less efficient
+- 👎 **Runtime** - Will require additional logic in `__styles()` (_artifact of AOT compilation in Griffel_) to prefix classes that makes the idea of AOT compilation obsolete
+
+## Current workaround
+
+We have proposed a temporary workaround in [microsoft/griffel#453](https://github.com/microsoft/griffel/issues/453#issuecomment-1850115159). The workaround involves increasing the specificity of the CSS rules in the consuming app. This is achieved by adding a unique class to the root element of the app, for example:
+
+```css
+.color-red {
+}
+/* becomes ⬇️ */
+.PREFIX .color-red {
+}
+```
+
+As a result, the CSS rules will have a higher specificity and will override any conflicting rules from the application. This approach is not ideal, as it affects performance (makes CSS rules less efficient).
```

Comment:
> it affects performance (makes CSS rules less efficient).

how much does it affect performance ? it's significant hit or minor ?

Reply:
Sadly, the answer is an unsatisfying, "it depends". Things the answer depends on are:

1. How many CSS selectors you have
2. How many DOM nodes you have
3. How many classes you have on an element
4. How complex are your CSS selectors (e.g., combinators, pseudos, etc)

In my testing in this area, the first two have the biggest impact, but from a Fluent perspective are what we have the least control over. The last certainly can impact performance, but has a smaller impact or tends to show up in edge cases, but is what we have the most control over in Fluent.

While I don't think we should outright ban "complex selectors" I feel that for Fluent we should tend toward sacrificing our convenience to leave perf on the table for consuming applications. So even if the performance hit is minor we should avoid it if at all possible.

IMO the bigger concern with this approach is called out in the RFC: it increases specificity making overriding somewhat harder.